### PR TITLE
Support variables in slice filter shorthand

### LINKED ISSUE #769: The slice filter shorthand doesn't support variables
The slice filter type regex is `regex: /^\[(\d*:\d*)\]/,` which only matches something like `[10:]` but not `[length:]`. I can just use `|slice(length)` and `|slice(0, length)` but it's a little unexpected coming from php twig.

### PR DESCRIPTION
**Pull request by antoineveldhoven**
Resolves #769

### CODE REVIEW

**Review by RobLoach — APPROVED** [review]
Oh this is nice. Thanks a lot. Looks good to me. Anything else we should address before merging?

**Comment by antoineveldhoven** [comment]
> Oh this is nice. Thanks a lot. Looks good to me. Anything else we should address before merging?

Thanks for the review, this commit should merge perfectly as-is.
Could you or @willrowe take a look at #874, we really need this to me merged for our project(s).

File: src/twig.expression.js [hunk 1]
```diff
@@ -576,14 +576,14 @@ module.exports = function (Twig) {
         },
         {
             type: Twig.expression.type.slice,
-            regex: /^\[(\d*:\d*)\]/,
+            regex: /^\[(-?\w*:-?\w*)\]/,
```
Comment (willrowe) [inline src/twig.expression.js:579]: It looks like you're adding support for negative values here, but there are no tests for this. Please add tests.

File: src/twig.expression.js [hunk 2]
```diff
@@ -596,11 +596,39 @@ module.exports = function (Twig) {
...
+            parse(token, stack, context) {
                 const input = stack.pop();
-                const {params} = token;
+                let {params} = token;
                 const state = this;
 
+                if (parseInt(params[0], 10).toString() === params[0]) {
+                    params[0] = parseInt(params[0], 10);
+                } else {
+                    const value = context[params[0]];
```
Comment (willrowe) [inline src/twig.expression.js:607]: This only allows very simple variable names. Technically PHP Twig supports full expressions as slice parameters, like accessing a member value of a map or a math expression. However, I'm just mention this to point it out. I'm fine with allowing this simple case to push things forward and don't thing it should be a blocker to get this merged.

File: test/test.expressions.js [hunk 3]
```diff
@@ -455,6 +455,12 @@ describe('Twig.js Expressions ->', function () {
             output.should.equal('23');
         });
 
+        it('should support slice shorthand (full form) with variables as arguments', function () {
```
Comment (willrowe) [inline test/test.expressions.js:458]: Please add more test cases to cover both the optional first and second parameters.

File: test/test.filters.js [hunk 4]
```diff
@@ -640,6 +640,10 @@ describe('Twig.js Filters ->', function () {
             const testTemplate = twig({data: '{{ \'12345\'|slice(1, 2) }}'});
             testTemplate.render().should.equal('23');
         });
```
Comment (willrowe) [inline test/test.filters.js:642]: Please add more test cases to cover both the optional first and second parameters.

**Review by willrowe — CHANGES_REQUESTED** [review]

File: src/twig.expression.js (hunk 2 above)
Comment (antoineveldhoven) [reply src/twig.expression.js:607]: Thanks for pointing this out!
I wasn't aware of it.

For now I think it is OK to allow these simple cases; I'll create an issue in order to support full expressions; but have to look into the current implementation in PHP before doing that.

**Comment by antoineveldhoven** [comment]
@willrowe Thanks for the extensive review; I will add the requested tests as soon as possible.

**Comment by antoineveldhoven** [comment]
@willrowe Can you take a look at my revision of this PR?
I;ve added an additional check for negative length parameter in the slice filter as in [PHP: array_slice](https://www.php.net/manual/en/function.array-slice.php )

**Review by willrowe — APPROVED** [review]

**Comment by JamesB797** [comment]
Nice, thank you for addressing this!

**Comment by willrowe** [comment]
@RobLoach can you go ahead and release v1.17.0? Everything in the milestone has been completed, so we might as well release instead of adding more at this point.

**Pull request merged**